node/src/node.cpp

```diff
@@ -44,19 +44,100 @@
 #include <boost/spirit/include/karma_list.hpp>
 #include <boost/spirit/include/karma_string.hpp>
 
+#include "cocaine/service/node/overseer.hpp"
+
 using namespace cocaine;
 using namespace cocaine::service;
 
+using cocaine::service::node::overseer_t;
+
 namespace ph = std::placeholders;
 
+class control_slot_t:
+    public io::basic_slot<io::node::control_app>
+{
+    struct controlling_slot_t:
+        public io::basic_slot<io::node::control_app>::dispatch_type
+    {
+        typedef io::basic_slot<io::node::control_app>::dispatch_type super;
+
+        typedef io::event_traits<io::node::control_app>::dispatch_type dispatch_type;
+        typedef io::protocol<dispatch_type>::scope protocol;
+
+        control_slot_t* p;
+        std::shared_ptr<overseer_t> overseer;
+
+        controlling_slot_t(const std::string& name, control_slot_t* p_):
+            super("controlling"),
+            p(p_)
+        {
+            overseer = p->parent.overseer(name);
+            if (overseer == nullptr) {
+                throw cocaine::error_t("!");
+            }
+
+            on<protocol::chunk>([&](int size) {
+                overseer->failover(size);
+            });
+
+            p->locked.store(true);
+        }
+
+        ~controlling_slot_t() {
+            p->locked.store(false);
+        }
+
+        virtual
+        void
+        discard(const std::error_code&) const {
+            overseer->failover(0);
+        }
+    };
+
+    typedef std::vector<hpack::header_t> meta_type;
+    typedef std::shared_ptr<const io::basic_slot<io::node::control_app>::dispatch_type> result_type;
+
+    std::atomic<bool> locked;
+    node_t& parent;
+
+public:
+    control_slot_t(node_t& parent):
+        locked(false),
+        parent(parent)
+    {}
+
+    boost::optional<result_type>
+    operator()(tuple_type&& args, upstream_type&& upstream) {
+        return operator()({}, std::move(args), std::move(upstream));
+    }
+
+    boost::optional<result_type>
+    operator()(const meta_type&, tuple_type&& args, upstream_type&& upstream) {
+        typedef io::protocol<io::event_traits<io::node::control_app>::upstream_type>::scope protocol;
+
+        const auto dispatch = tuple::invoke(std::move(args), [&](const std::string& name) -> result_type {
+            if (locked) {
+                upstream.send<protocol::error>(std::make_error_code(std::errc::resource_unavailable_try_again));
+                return nullptr;
```

Comment:
Strictly speaking - this is an error(protocol violation) because rx dispatch is not closed. Only tx dispatch is closed with an error.
But it is not a big deal from my point of view.